Position scaled html within available container space

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Fixes https://github.com/WordPress/gutenberg/issues/65188; fixes https://github.com/WordPress/gutenberg/issues/65595

## What?
<!-- In a few words, what is the PR actually doing? -->
Rework how the scaled canvas is centered to fix a few bugs.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
- iframe scrollbar should be available. The previous implementation pushed the iframe too far to the right, meaning the [iframe scrollbar was hidden](https://github.com/WordPress/gutenberg/issues/65595).
- vertical toolbar and inserters were not updating their position when the editor chrome changed (sidebars opening/closing)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Affix the iframe to the right side of its container so the scrollbar is always visible
- Center the scaled html canvas by using `translateX` to account for open sidebars.

The gist is, rather than push the iframe around, push the scaled html element around.

Before: Margin-left to move the center of the iframe to line up with the container
After: Iframe takes up full available space, scale html, then push it over to be in the available area

## Known Reflow Points
There will be some reflow when:
- Start with a sidebar open
- Enter zoom out
- Close the sidebar

This is because our initial scale point started from a size smaller than our total available window space (all sidebars closed). In this one instance, I think reflow is acceptable. If we don't reflow, then it reflows when we exit zoom out, which I think is worse. 

Most important points for reflow to NOT happen:
- When entering zoom out
- When exiting zoom out

When it could be allowed:
- When opening/closing sidebars

One way we can prevent this is _if beginning from a smaller state_ (sidebar open, enter zoom out) then add padding on the iframe to force the HTML to account for the extra space and stay scaled smaller. I'm not sure this complexity is worth it though.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

**Iframe before.**
Note the left side is pushed over. This was to center the scaled canvas in the container. It worked to center the canvas, but the scrollbar is also pushed to the right. 
<img width="1488" alt="screenshot of iframe being pushed to the right" src="https://github.com/user-attachments/assets/a1a2399e-9482-4bce-b932-6cad68e55912">


**iframe after**
The iframe is the same size as the previous container. This allows the scrollbar to be visible. Then, the scaled content is pushed to the left to take up its available space. 
<img width="1487" alt="iframe takes up entire space" src="https://github.com/user-attachments/assets/a42a4f29-9da9-4b09-b824-3dc04fe82e0c">

jeryj committed:
Position scaled html within available container space

jeryj commented on packages/block-editor/src/components/iframe/style.scss:
```diff
@@ -9,9 +9,10 @@
-	$container-width: var(--wp-block-editor-iframe-zoom-out-container-width, 100vw);
 	$prev-container-width: var(--wp-block-editor-iframe-zoom-out-prev-container-width, 100vw);
 	width: $prev-container-width;
-	// This is to offset the movement of the iframe when we open sidebars
-	margin-left: calc(-1 * (#{$prev-container-width} - #{$container-width}) / 2);
```
This was causing the scrollbar to be hidden off canvas.

jeryj commented on packages/block-editor/src/components/iframe/style.scss:
```diff
@@ -9,9 +9,10 @@
-	margin-left: calc(-1 * (#{$prev-container-width} - #{$container-width}) / 2);
+	// Position the iframe so that it is always aligned with the right side so that
+	// the scrollbar is always visible on the right side
+	position: absolute;
+	right: 0;
```
Affix the iframe container to the right side of the container so its scrollbar is stuck to the right side.

ciampo replied:
Will this work well with RTL writing direction?

MaggieCabrera replied:
would adding `/*rtl:ignore*/` before line 17 be enough?

ciampo replied:
I think @draganescu already tested it in an RTL environment and [said it's looking fine](https://github.com/WordPress/gutenberg/pull/66034#issuecomment-2407011122) :)

jeryj commented on packages/block-editor/src/components/iframe/content.scss:
```diff
@@ -13,9 +13,9 @@
-
-	transform: scale(#{$scale});
-
+	$container-width: var(--wp-block-editor-iframe-zoom-out-container-width, 100vw);
+	// Apply an X translation to center the scaled content within the available space.
+	transform: scale(#{$scale}) translateX(calc(( #{$prev-container-width} - #{ $container-width }) / 2 / #{$scale}));
```
Move the scaled canvas to the right to occupy the space of the container.

stokesman replied:
I think the `translateX()` could precede the `scale()` and the translation wouldn’t have to be manually counter-scaled but that’s a nit.

ciampo replied:
We could also use individual [`scale`](https://developer.mozilla.org/en-US/docs/Web/CSS/scale) and [`translate`](https://developer.mozilla.org/en-US/docs/Web/CSS/translate) properties — if that helps changing them independently without rewriting the whole `transform`

jeryj replied:
> I think the translateX() could precede the scale()

We tried that as well, but it seems like it's still necessary to apply the scaling to the `translateX`. 

> We could also use individual scale and translate properties

I'm going to try this, as I think we'll need to adjust the animating timings for this separately.

ajlende commented on packages/block-editor/src/components/iframe/style.scss:
```diff
@@ -9,9 +9,10 @@
 }
 
 .block-editor-iframe__scale-container.is-zoomed-out {
-	$container-width: var(--wp-block-editor-iframe-zoom-out-container-width, 100vw);
```
We should be able to remove this from the JS too since it's no longer used outside the iframe.

stokesman reviewed (commented):
This tests well for me. An added bonus is that it fixes the centering of the canvas on devices whose scrollbars are shown. I noticed there can be an odd transition when exiting zoom out:

https://github.com/user-attachments/assets/ef67e6b3-2c3a-491a-835f-c058a44940d2

I don’t think it’s a big deal considering all the other odd things that can go on during the transition and this already seems to improve some of those.

Great idea on the approach. I had the same (rough) idea in the back of my mind and I’m glad to see it seemed simple enough.

ciampo reviewed (commented):
Gave this a look, and it seems to work much better!

The timing of the animation can be improved, but I'm sure that it can be worked out.

I also found a bug (~36 seconds in the video below):

- open a teamplate
- enable zoomed out view 
- click on the W logo to open side editor sidebar
- click on another template
- zoomed out view is already enabled, and the iframe is misplaced

https://github.com/user-attachments/assets/46e9371c-19d9-48ce-8025-b5fa0d4e5152


I'll let other folks with a lot more familiarity on this part of Gutenberg review this PR in depth (@draganescu @MaggieCabrera @getdave )

getdave commented:
Based on @draganescu's review this feels close. 

@jeryj can you let me know if this PR becomes blocked or gets very much more complex? I just want to keep an eye on the scope of the fixes targeting 6.7 (which I assume this is?).

jeryj committed:
Move scale to own css line

MaggieCabrera commented:
This is very close, and it doesn't change the size of the canvas when there are more or less sidebars open. Great job here!

MaggieCabrera commented:
> Gave this a look, and it seems to work much better!
> 
> The timing of the animation can be improved, but I'm sure that it can be worked out.
> 
> I also found a bug (~36 seconds in the video below):
> 
>     * open a teamplate
> 
>     * enable zoomed out view
> 
>     * click on the W logo to open side editor sidebar
> 
>     * click on another template
> 
>     * zoomed out view is already enabled, and the iframe is misplaced
> 

Do you think this is similar to https://github.com/WordPress/gutenberg/issues/65783?

ciampo commented:
> > Gave this a look, and it seems to work much better!
> > The timing of the animation can be improved, but I'm sure that it can be worked out.
> > I also found a bug (~36 seconds in the video below):
> > ```
> > * open a teamplate
> > 
> > * enable zoomed out view
> > 
> > * click on the W logo to open side editor sidebar
> > 
> > * click on another template
> > 
> > * zoomed out view is already enabled, and the iframe is misplaced
> > ```
> 
> Do you think this is similar to #65783?

The "zoomed out mode is already enabled" part is the same as #65783, but the other part of the bug that I wanted to highlight is that the iframe/canvas is clipped and misplaced when it loads in zoomed-out mode

jeryj committed:
Always have width of scale container be full width avaialable

The scale container could be smaller than the available area if zoom out was initialized with a sidebar open, then closed. We were using prevContainerWidth in the CSS under the assumption it would always be the largest value, but that was not true.

This renames prevContainerWidthRef to initialContainerWidth so it's clearer that this is the point when zoom out was initialized, then renames the CSS vars prev-container-width to outer-container-width, using the larger value of containerWidth or initialContainerWidth so it can be more consistently named.

jeryj committed:
Comment about bug

ajlende committed:
Remove unused css var

jeryj committed:
Force largest available window size to scale from

If you started with a sidebar open, then entered zoom out, then closed the sidebar, your scaled canvas would be too large. It should match the same size as if you start with no sidebars open, then enter zoom out.

jeryj committed:
Improve animation by not transitioning the translateX

jeryj commented on packages/block-editor/src/components/block-canvas/style.scss:
```diff
@@ -4,5 +4,4 @@ iframe[name="editor-canvas"] {
 	height: 100%;
 	display: block;
 	background-color: transparent;
-	@include editor-canvas-resize-animation;
```
This doesn't appear to have any effect.

jeryj committed:
Remove unused comment

jeryj commented:
EOD Notes:

There's one more way to improve this that I couldn't dial in: We only want to animate the scaling when _entering_ or _exiting_ zoom out. We don't want to animate scale when it updates from sidebars opening. This is what causes it to fly out to random areas. To reproduce:
- Enter zoom out
- Open Right sidebar
- Open left sidebar

If we don't animate scale at that point, it's handled by the resizing of the iframe window and looks much smoother. @MaggieCabrera Do you have any ideas on how to do that efficiently?

MaggieCabrera commented:
> If we don't animate scale at that point, it's handled by the resizing of the iframe window and looks much smoother. @MaggieCabrera Do you have any ideas on how to do that efficiently?
> 

I need to test this more thoroughly but my first thought is that we are animating everything that happens to the div, maybe we want to target something specific like the scale and that would be enough? As a CSS only solution my only other suggestion is to have a class specific to the moments we want animated and then remove it, but I bet that's something that framer does on its own, so maybe that's another avenue to explore

draganescu commented on packages/block-editor/src/components/iframe/content.scss:
```diff
@@ -5,18 +5,20 @@
 .block-editor-iframe__html {
 	transform-origin: top center;
-	@include editor-canvas-resize-animation;
+	transition: all 0.4s cubic-bezier(0.46, 0.03, 0.52, 0.96), transform 0s;
```
We could use a comment about the numbers even like, "this is what works" 😄

jeryj replied:
It was just removing the mixin since it's only used once: https://github.com/WordPress/gutenberg/pull/66034/files#diff-3014a931b478b662a3c554c1306e3b548da2c2edd5b32114395b329404f8f50aL40

draganescu reviewed (approved):
This tests great for me with the rebase and the zoom out hook fixed - I can't repro the clipping which was the biggest issue of this PR. The bouncing is greatly reduced too

getdave reviewed (dismissed):
I tried every which way I could to break this but it seems really solid. Great work everyone.

jeryj committed:
Only animate scaling on entry and exit of zoom out mode

jeryj committed:
Use max of initialContainerWidth or containerWidth to avoid scaling canvas over 1

getdave reviewed (commented):
The key aspects of this PR are working well and it resolves the original bug which was vertical toolbars being left "unpositioned" when toggling sidebars.

After updating to the latest changes since my last review, I still notice that we have animations happening on inserters and toolbars when toggling sidebars. These look a little funky but shouldn't block this PR merging.

<details>
<summary>Animations on toggle sidebars</summary>

https://github.com/user-attachments/assets/1e3d7b49-9fbe-47c3-bdb4-69ca8b64f6b8

</details>

I recommend we merge this PR to improve the stability of the feature and then follow up to try and resolve the sidebar animations.

@kevin940726 @draganescu how do you feel about this and targetting 6.7?

jeryj committed:
Allow zoom out scale e2e test to be more flexible

jeryj commented:
@dhruvang21 Thank you for testing! I went ahead and merged this as the issues you found are both present on `trunk`.

1. The [patterns tab entering zoom out mode](https://github.com/WordPress/gutenberg/pull/65785) is intentional
2. The multiple scrollbars on the preview is also present on trunk (and should be fixed separately). I wasn't aware of this bug. Thanks for finding it!

getdave commented:
@jeryj Could you take point on ensuring this PR cherry picks cleanly to the `wp/6.7` branch. Note it may be this PR is relying on other PRs which have not been backported. @draganescu had a similar issue.

ajlende commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -336,13 +374,16 @@ function Iframe( {
 
-		return () => {
-			iframeDocument.documentElement.classList.remove( 'is-zoomed-out' );
+		// iframeDocument.documentElement.style.setProperty(
+		// 	'--wp-block-editor-iframe-zoom-out-outer-container-width',
+		// 	`${ Math.max( initialContainerWidth.current, containerWidth ) }px`
+		// );
```
I don't think these comments were meant to be committed.

jeryj replied:
Nope!

ajlende commented on test/e2e/specs/site-editor/zoom-out.spec.js:
```diff
@@ -27,8 +27,8 @@ test.describe( 'Zoom Out', () => {
-			'transform',
-			'matrix(0.67, 0, 0, 0.67, 0, 0)'
+			'scale',
+			new RegExp( /0\.[5-8][0-9]*/, 'i' )
```
Why `0.5` to `0.89`? And why not use `parseInt` to check the range instead of `RegExp`?

jeryj replied:
[toHaveCSS](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-have-css) only accepts string or regex. And I did a range so it would allow us to update the the zoom level scaling or sizing without this failing. .5-0.89 seemed like a reasonable range.

ajlende commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -298,14 +298,49 @@ function Iframe( {
 	useEffect( () => cleanup, [ cleanup ] );
 
+	const zoomOutAnimationClassnameRef = useRef( null );
+	const handleZoomOutAnimationClassname = () => {
+		clearTimeout( zoomOutAnimationClassnameRef.current );
+
+		iframeDocument.documentElement.classList.add( 'zoom-out-animation' );
+
+		zoomOutAnimationClassnameRef.current = setTimeout( () => {
+			iframeDocument.documentElement.classList.remove(
+				'zoom-out-animation'
+			);
+		}, 400 ); // 400ms should match the animation speed used in components/iframe/content.scss
+	};
```
This function should be defined inside the `useEffect` since that's the only place it's used.

ajlende commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -336,13 +374,16 @@ function Iframe( {
-			'--wp-block-editor-iframe-zoom-out-prev-container-width',
-			`${ prevContainerWidthRef.current }px`
+			'--wp-block-editor-iframe-zoom-out-outer-container-width',
+			`${ Math.max( initialContainerWidth.current, containerWidth ) }px`
```
I would make this a variable since it's used in three places where they are expected to always be the same.

ajlende commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -336,13 +374,16 @@ function Iframe( {
 		iframeDocument.documentElement.style.setProperty(
-			'--wp-block-editor-iframe-zoom-out-prev-container-width',
-			`${ prevContainerWidthRef.current }px`
+			'--wp-block-editor-iframe-zoom-out-outer-container-width',
```
I would maybe rename this to `--wp-block-editor-iframe-zoom-out-scale-container-width` instead since this is the width of the `block-editor-iframe__scale-container`.

jeryj replied:
We may not always call it scale container, so I'd be fine leaving this one named as is.

ajlende replied:
Right, it may be changed, but for the current reader, the connection between the two would be more clear if the variable was named the same as the container that it's setting the width for.

ajlende replied:
`--wp-block-editor-iframe-scale-container-width` would be acceptable too.

getdave commented:
@ajlende @jeryj As I understand it this got backported as part of https://github.com/WordPress/gutenberg/pull/66182 so I'm updating the labels accordingly.
